### PR TITLE
Feature/custom checkers unit tests

### PR DESCRIPTION
**Pull request by albalabkin**

### CODE REVIEW

File: common/_rest_qa_api/response_validator.py [hunk 1]
```diff
@@ -197,3 +197,4 @@ def _run_checkers(self: Union['BaseResponseModel', 'ResponseValidatorMixin']):
                 raise
             else:
                 logger.info(f"{checker.__name__} validation passed")
+            # TODO - 'validation passed' message triggered in any case
```
Comment (maximrodin) [inline common/_rest_qa_api/response_validator.py]: Could you please fix it in scope of this PR?

File: common/hooks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pytest_plugins = [
+    "common.hooks.cli_options",
+    "common.hooks.logger_hooks",
+    "common.hooks.pycats_hooks"
+]
```
Comment (maximrodin) [inline common/hooks/__init__.py]: Add new line please

File: conftest.py [hunk 3]
```diff
@@ -1,3 +1,3 @@
 pytest_plugins = [
-    "common.hooks.pycats_hooks"
+    "common.hooks"
 ]
```
Comment (maximrodin) [inline conftest.py]: Add new line please

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+config = DummyConfigBuilder(DummyApiValidationConfig())
+
+
+def fake_response():
+    return DummyResponseBuilder().method().code().header({}).body()
+
+@patch('requests.request', return_value=fake_response())
+class TestRunCheckersBranches:
+   def test_BaseRESTCheckers_instances_for_customer_checker(self, _, response, builder):
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: Please, use lower case in functions/methods/tests names

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        with pytest.raises(TypeError) as excinfo:
+            builder.endpoint.get()
+        assert f"{DummyResponseBuilder} has unsupported type. Supported are " \
+               f"functions and BaseRESTCheckers instances" in str(excinfo.value)
+
+   def test_exceptions_for_customer_checkers(self, _, response, builder, caplog):
+        builder.endpoint.response_model.custom_checkers.append(DummyResponseBuilder)
+        with pytest.raises(Exception):
+            builder.endpoint.get()
+        expected_log_message = f"fail"
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py:64]: _run_checkers method logs exception with error right after fail. Please add this check to test

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+@patch('requests.request', return_value=fake_response())
+class TestRunCheckersBranches:
+   def test_BaseRESTCheckers_instances_for_customer_checker(self, _, response, builder):
+        JSONCheckers.status = 400
+        JSONCheckers.deactivate(JSONCheckers.check_json_structure)
+        builder.endpoint.response_model.custom_checkers.append(JSONCheckers)
+        response.status_code = 200
+        with pytest.raises(RestResponseValidationError) as excinfo:
+            builder.endpoint.get()
+        assert f"Invalid status code. Expected '{JSONCheckers.status}', " \
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: Missing message part: "Checker 'check_status' failed with error: "

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        builder.endpoint.response_model.custom_checkers.append(check_status(check_status_code))
+        try:
+            builder.endpoint.get()
+        except Exception as e:
+            pytest.fail(f"DID RAISE {e}")
+        expected_log_message = f"{JSONCheckers.check_status.__name__} validation passed"
+        messages = [record for record in caplog.records if expected_log_message == record.message]
+        assert len(messages) == 1, "Expected message not found in logs"
+
+    def test_function_call(self, _, response, builder):
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: Why it's in TestCustomLogger class?

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        response.status_code = 200
+        try:
+            builder.endpoint.get()
+        except Exception as e:
+            pytest.fail(f"DID RAISE {e}")
+        expected_log_message = f"{JSONCheckers.check_status.__name__} validation passed"
+        messages = [record for record in caplog.records if expected_log_message == record.message]
+        assert len(messages) == 1, "Expected message not found in logs"
+
+    def test_logging_through_customer_checker_class(self, _, response, builder, caplog):
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: What is the difference with the previous test?

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            builder.endpoint.get()
+        expected_log_message = f"fail"
+        messages = [record for record in caplog.records if expected_log_message == record.message]
+        assert len(messages) == 1, "Expected message not found in logs"
+
+
+
+@patch('requests.request', return_value=fake_response())
+class TestCustomLogger:
+    def test_logging_after_successful_execution_of_customer_checker(self, _, response, builder, caplog):
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: custom, not customer

Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: add an empty line between class and method name

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
+from unittest.mock import patch
+
+import pytest
+from common._rest_qa_api.rest_exceptions import  RestResponseValidationError
+from common._rest_qa_api.rest_checkers import JSONCheckers, check_status
+from unit_tests.rest_qa_api_tests.tests_utils import DummyApiValidationConfig, DummyConfigBuilder, DummyResponseBuilder
+
+config = DummyConfigBuilder(DummyApiValidationConfig())
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: Why do you need config here?

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+from unittest.mock import patch
+
+import pytest
+from common._rest_qa_api.rest_exceptions import  RestResponseValidationError
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: 2 spaces after import

File: unit_tests/rest_qa_api_tests/custom_checkers_test.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        JSONCheckers.deactivate(JSONCheckers.check_json_structure)
+        builder.endpoint.response_model.custom_checkers.append(JSONCheckers)
+        response.status_code = 200
+        with pytest.raises(RestResponseValidationError) as excinfo:
+            builder.endpoint.get()
+        assert f"Invalid status code. Expected '{JSONCheckers.status}', " \
+               f"but got '{response.status_code}'" in str(excinfo.value)
+
+   def test_function_call_and_assertion_branch_for_customer_checkers(self, _, response, builder):
+        response.status_code = 200
```
Comment (maximrodin) [inline unit_tests/rest_qa_api_tests/custom_checkers_test.py]: JSONCheckers class will have status code 400 after the first test - it's bad practice because the test data are not atomic. 
Please create a fixture that will reset JSONCheckers state before running the tests

**Review by maximrodin — CHANGES_REQUESTED** [review]
Please review unused files .coverage, selenium log lock.
Check indent in custom_checkers_test.py - looks like it over-indented.
Add tests for multiple checkers executed - _run_checkers execute checkers in the loop.
Add tests for error message format for custom checkers (placeholders ####,etc.)

File: common/_rest_qa_api/response_validator.py [hunk 13]
```diff
@@ -196,5 +196,5 @@ def _run_checkers(self: Union['BaseResponseModel', 'ResponseValidatorMixin']):
                 logger.exception(err)
                 raise
             else:
-                logger.info(f"{checker.__name__} validation passed")
-            # TODO - 'validation passed' message triggered in any case
\ No newline at end of file
+                if not self.errors["default"] and not self.errors["custom"]:
+                    logger.info(f"{checker.__name__} validation passed")
```
Comment (maximrodin) [inline common/_rest_qa_api/response_validator.py:200]: Add unit test for negative condition

**Review by maximrodin — COMMENTED** [review]
Remove .coverage file

**Review by maximrodin — APPROVED** [review]
